### PR TITLE
Reduce laser pointer lag

### PR DESCRIPTION
**Pull request by SamGondelman**
Addresses fogbugz [3842](https://highfidelity.fogbugz.com/f/cases/3842/Reduce-Laser-Pointer-Lag)

Lots of scripts require "laser pointers" and were all creating, managing, and updating their own overlays to visualize them.  Since the scripts were updating at most at 60 fps, these laser pointers would demonstrate significant lag compared to everything else.  This PR abstracts the logic required for laser pointers on the C++ side, which means they update and render as fast as everything else and no longer lag.  Additionally, if multiple scripts create identical laser pointers (same origin/direction), now only one intersection test is performed, which should further improve performance.

Laser pointers can either update based on a joint (with local position and direction offsets), stay at a set position/direction, or follow the mouse.  You can also set the following properties:
- filter: filters on the types of intersections (Entities, Overlays, Avatars, HUD, include invisible, include noncollidable, possibly more to come later)
- maxDistance: a max distance for the ray to go
- faceAvatar: if the end overlay should always face your avatar or not
- centerEndY: if the end overlay should be centered at the intersection or moved up by half its height
- lockEnd: if the end of the overlay should lock on to the position of its target
- renderStates: a list of "render states" that define how to graphically represent the laser pointer in different cases
- defaultRenderStates: a list of render states to use when no intersection is detected.  Must contain an additional "distance" field to define the distance at which the end is drawn.

To use these tools there are two new JS APIs:
- RayPick JS API
    - bitmask values for all the different filter options (RayPick.PICK_ENTITIES, RayPick.PICK_OVERLAYS, RayPick.PICK_AVATARS, RayPick.PICK_HUD, RayPick.INCLUDE_INVISIBLE, RayPick.INCLUDE_NONCOLLIDABLE)
    - values for all the different intersection types (RayPick.INTERSECTED_NONE, RayPick.INTERSECTED_ENTITY, RayPick.INTERSECTED_OVERLAY, RayPick.INTERSECTED_AVATARS, RayPick.INTERSECTED_HUD)
    - createRayPick
    - enableRayPick
    - disableRayPick (disabled ray picks don't perform intersection tests)
    - removeRayPick
- LaserPointers JS API
    - createLaserPointer
    - enableLaserPointer
    - disableLaserPointer (disabled laser pointers don't render or perform intersection tests)
    - removeLaserPointer
    - setRenderState (allows you to switch between user-defined visual representations)
    - editRenderState (allows you to modify a specific render state)
    - setLockEndUUID (locks laser pointer onto specific overlay or entity)
- Both:
    - getPrevRayPickResult (tells you type of last intersection (type), ID of intersected entity/avatar/overlay (objectID), distance to intersection (distance), intersection point (intersection), and surface normal at intersection (surfaceNormal))
    - set{Ignore/Include}{Entities/Overlays/Avatars} (set list of IDs to include/ignore)

There's also a new Overlay property, drawHUDLayer, which enables you to draw overlays in front of the HUD.  Overlays drawn in this layer will not be lit, and will not show up behind semi-transparent HUD elements.  Otherwise they will depth test against each other and the HUD sphere.  HUD layer overlays will also not be rendered at all in non-HMD modes.

Also, converts several scripts to use the new APIs:
- teleport.js
- grab.js (slight difference: white beacon line is now a thin cube)
- handControllerPointer.js (slight difference in appearance of laser)
- handControllerGrab.js (slight difference in appearance of laser)

[Here's](https://gist.githubusercontent.com/SamGondelman/2ca75f058bc7807b4dd3ddd38065370a/raw/adc07fd851e899b3ca350285f8b9bd465ff5cca0/laserPointerTest.js) a test script you can run that shows off pretty much all of the functionality.  Use 5 and 6 to cycle through a bunch of example laser pointers with different properties, and 1-4 to switch render states.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1721/HighFidelity-Beta-PR10980-197ba83b103ab1caf74d2905b8936d87df8e838d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1721/HighFidelity-Beta-PR10980-197ba83b103ab1caf74d2905b8936d87df8e838d.exe

**Comment by PhilipRosedale** [comment]
add to whitelist

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1729/HighFidelity-Beta-PR10980-0c8c6ff5d677abe7b9ba5792920a9c286ec3702d.exe

**Comment by SamGondelman** [comment]
Profile from master showing time between teleport.js updates (running at ~60 fps):
teleport_test_master


Profile from this PR showing time between updating all laser pointers (at the update framerate, ~90 fps):
teleport_test_pr

**Comment by SamGondelman** [comment]
test this please

**Comment by SamGondelman** [comment]
Test plan:

- On release, using the old teleport.js, observe that in areas with reasonable amounts of content (dev-welcome works fairly well), the teleport beams noticeably lag behind your hands.  In particular, the origin of the ray doesn't follow your hand very smoothly as you move your hand around quickly.
- With this PR and the new teleport.js, the teleport beams should follow your hands exactly, regardless of how much content is around.
- Otherwise, the new teleport.js should still behave the same as before (check invisible entities, visible entities, and seats).  One change: laser beams will no longer appear if they aren't intersecting anything (possibly needs to be fixed).
- Take a look at the description at the top of the [test script](https://gist.githubusercontent.com/SamGondelman/2ca75f058bc7807b4dd3ddd38065370a/raw/5564227234aa99726bd1455a62c1f74dc558cd92/laserPointerTest.js).  Run it and confirm that all the lasers with their various settings behave as you'd expect.  You can look at the definitions for each laser pointer to get a better sense of what they should each do.  Hopefully it's self-explanatory but I can also answer any questions or add better descriptions to the script if that would help.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1736/HighFidelity-Beta-PR10980-0c8c6ff5d677abe7b9ba5792920a9c286ec3702d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1736/HighFidelity-Beta-PR10980-0c8c6ff5d677abe7b9ba5792920a9c286ec3702d.exe

**Comment by PhilipRosedale** [comment]
Seems very fast to me, at first test, bravo!  

I get a problem where the tablet/UI lasers sometimes get moved to emitting above my hand.  Clicking the regular seek lasers against the scene fixes it.

File: interface/src/raypick/LaserPointerManager.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+//  Copyright 2017 High Fidelity, Inc.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+#include "LaserPointerManager.h"
+#include "LaserPointer.h"
+#include "RayPick.h"
+
+LaserPointerManager& LaserPointerManager::getInstance() {
```
Comment (ZappoMan) [inline interface/src/raypick/LaserPointerManager.cpp]: use DependencyManager

File: interface/src/raypick/LaserPointerManager.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+#include <glm/glm.hpp>
+
+#include "LaserPointer.h"
+
+class RayPickResult;
+
+class LaserPointerManager {
+
+public:
+    static LaserPointerManager& getInstance();
```
Comment (ZappoMan) [inline interface/src/raypick/LaserPointerManager.h]: use DependencyManager

File: interface/src/raypick/LaserPointerScriptingInterface.cpp [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+#include "LaserPointerScriptingInterface.h"
+
+#include <QVariant>
+#include "RegisteredMetaTypes.h"
+#include "GLMHelpers.h"
+
+#include "Application.h"
+
+LaserPointerScriptingInterface* LaserPointerScriptingInterface::getInstance() {
```
Comment (ZappoMan) [inline interface/src/raypick/LaserPointerScriptingInterface.cpp]: use DependencyManager

File: interface/src/raypick/LaserPointerScriptingInterface.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+#include <QtCore/QObject>
+
+#include "LaserPointerManager.h"
+#include "RegisteredMetaTypes.h"
+
+class LaserPointerScriptingInterface : public QObject {
+    Q_OBJECT
+
+public:
+    static LaserPointerScriptingInterface* getInstance();
```
Comment (ZappoMan) [inline interface/src/raypick/LaserPointerScriptingInterface.h]: use DependencyManager

File: interface/src/raypick/RayPickManager.cpp [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            rayPick->setRayPickResult(RayPickResult());
+        }
+    }
+}
+
+unsigned int RayPickManager::addRayPick(std::shared_ptr<RayPick> rayPick) {
+    // TODO:
+    // use lock and defer adding to prevent issues
+    _rayPicks[_nextUID] = rayPick;
+    return _nextUID++;
```
Comment (SamGondelman) [inline interface/src/raypick/RayPickManager.cpp]: I could switch to QUuids.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1755/HighFidelity-Beta-PR10980-55a025b827ab5184e3d1311bf225f1004d0f25ee.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1755/HighFidelity-Beta-PR10980-55a025b827ab5184e3d1311bf225f1004d0f25ee.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1790/HighFidelity-Beta-PR10980-bf243d60255c8ce940ae86ef9f59e8bba7493bc7.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1790/HighFidelity-Beta-PR10980-bf243d60255c8ce940ae86ef9f59e8bba7493bc7.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1823/HighFidelity-Beta-PR10980-c7f0f03fcebdc30782d04559787004f5996f197c.exe

**Comment by SamGondelman** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1829/HighFidelity-Beta-PR10980-c7f0f03fcebdc30782d04559787004f5996f197c.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1910/HighFidelity-Beta-PR10980-3c8b964cc9d10cd4e50c9672a0941118303dd497.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1910/HighFidelity-Beta-PR10980-3c8b964cc9d10cd4e50c9672a0941118303dd497.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1938/HighFidelity-Beta-PR10980-3149c3e16c0671a8428390b28e3614f3ca170c90.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1938/HighFidelity-Beta-PR10980-3149c3e16c0671a8428390b28e3614f3ca170c90.exe

**Comment by SamGondelman** [comment]
This is a busy image but you should be able to get a sense of how drawHUDLayer works:
drawhudlayer example

When you click the Running Scripts window so that it has focus this strange oval appears, but I believe that's a separate issue.  You can also see that the mouse pointer is drawn on top:
drawhudlayer bug or qml issue

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1958/HighFidelity-Beta-PR10980-658acd91f2fdb30be4b134869294c3094f9d58f3.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1958/HighFidelity-Beta-PR10980-658acd91f2fdb30be4b134869294c3094f9d58f3.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1960/HighFidelity-Beta-PR10980-3a8ad8b6318e629826d341df05a52da5f217dc0a.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1960/HighFidelity-Beta-PR10980-3a8ad8b6318e629826d341df05a52da5f217dc0a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1966/HighFidelity-Beta-PR10980-6e78238f588684da0ffe3a7f1a32d86bbec35bb1.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1966/HighFidelity-Beta-PR10980-6e78238f588684da0ffe3a7f1a32d86bbec35bb1.dmg

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2008/HighFidelity-Beta-PR10980-fcaa39b245eeeb9fb98939acb8953298ebdb05ec.dmg

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2025/HighFidelity-Beta-PR10980-fcaa39b245eeeb9fb98939acb8953298ebdb05ec.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows2016/2025/HighFidelity-Beta-PR10980-fcaa39b245eeeb9fb98939acb8953298ebdb05ec.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2039/HighFidelity-Beta-PR10980-e3288a9004af5ce42d4538490ffd9fb7390426e8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows2016/2039/HighFidelity-Beta-PR10980-e3288a9004af5ce42d4538490ffd9fb7390426e8.exe

**Comment by SamGondelman** [comment]
build this please

File: interface/resources/shaders/hmd_ui.frag [hunk 6]
```diff
@@ -0,0 +1,30 @@
+//
+//  Created by Bradley Austin Davis on 2016/07/11
```
Comment (jherico) [inline interface/resources/shaders/hmd_ui.frag]: I didn't write this

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2041/HighFidelity-Beta-PR10980-e3288a9004af5ce42d4538490ffd9fb7390426e8.dmg

File: interface/resources/shaders/hmd_ui.frag (hunk 6 above)
Comment (SamGondelman) [inline interface/resources/shaders/hmd_ui.frag]: it's hmd_ui_glow.frag with all the hand laser stuff removed which confused git but technically yes you did

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows2016/2041/HighFidelity-Beta-PR10980-e3288a9004af5ce42d4538490ffd9fb7390426e8.exe

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2052/HighFidelity-Beta-PR10980-e3288a9004af5ce42d4538490ffd9fb7390426e8.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows2016/2052/HighFidelity-Beta-PR10980-e3288a9004af5ce42d4538490ffd9fb7390426e8.exe

**Comment by Nex-Pro** [comment]
Taking a look at this.

File: interface/resources/shaders/hmd_ui.frag (hunk 6 above)
Comment (jherico) [inline interface/resources/shaders/hmd_ui.frag]: http://i2.kym-cdn.com/photos/images/newsfeed/001/079/173/ed2.png

**Comment by Nex-Pro** [comment]
## FAIL

I am seeing fast and smooth teleport beams.
However this has changed the laser pointer behaviour.
During the tutorial I noticed that the laser would project beyond the target object ( trigger held ) 
I created a small dynamic cube and dragged it around. Same thing happened. Also the dynamic ents lag behind the lasers.  The lasers don't look to be in contact with the object that is being manipulated.

Lasering onto a dialog shows a sharp edged circle at the contact point where in latest dev build and previously we get a soft edged contact spot.

Got a reproducible crash in tutorial using the teleport beams at /0,0,0 immediately after the rotate segment of the tutorial.

to repro. Swish the teleport beam rapidly to the left and back so it goes effectively behhind your avatar. 
This  crashed me  4 times. I could not invoke this crash in latest dev build #6945

**Comment by Nex-Pro** [comment]
Will be re-checking this later as requested.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2284/HighFidelity-Beta-PR10980-8b1c24e636f38793ee7dca1442e55ee4f458a940.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2284/HighFidelity-Beta-PR10980-8b1c24e636f38793ee7dca1442e55ee4f458a940.exe

**Comment by Nex-Pro** [comment]
Retesting

**Comment by Nex-Pro** [comment]
### FAIL

Crashes when swishing tp beam around as per repro above.

Unhandled exception at 0x00007FFE75D08353 (ntdll.dll) in interface.exe: 0xC0000374: A heap has been corrupted (parameters: 0x00007FFE75D5F6B0).

Will try to get a stack dump later. ( can't now get  thru the tutorial to the rotate part ) Will do another fresh install and debug.

**Comment by Nex-Pro** [comment]
Unhandled exception at 0x00007FFE40ACB2A0 (Qt5Script.dll) in interface.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF.

>	Qt5Script.dll!00007ffe40acb2a0()	Unknown
 	Qt5Script.dll!00007ffe40a62a22()	Unknown
 	Qt5Script.dll!00007ffe40a617a9()	Unknown
 	Qt5Script.dll!00007ffe40acc3bb()	Unknown
 	interface.exe!00007ff6ebbc0688()	Unknown
 	interface.exe!00007ff6ebbb2c90()	Unknown
 	Qt5Script.dll!00007ffe40ac293a()	Unknown
 	Qt5Script.dll!00007ffe40ae593d()	Unknown
 	Qt5Script.dll!00007ffe40ae1c36()	Unknown
 	Qt5Script.dll!00007ffe40ae8220()	Unknown
 	Qt5Script.dll!00007ffe40ae6360()	Unknown
 	Qt5Script.dll!00007ffe40a60f1a()	Unknown
 	Qt5Script.dll!00007ffe40a4cd84()	Unknown
 	Qt5Script.dll!00007ffe40a46e68()	Unknown
 	Qt5Script.dll!00007ffe40a76dd2()	Unknown
 	Qt5Script.dll!00007ffe40a60faf()	Unknown
 	Qt5Script.dll!00007ffe40ae7ede()	Unknown
 	Qt5Script.dll!00007ffe40aee3ae()	Unknown
 	Qt5Core.dll!0000000072e77db1()	Unknown
 	interface.exe!00007ff6ebca32e5()	Unknown
 	interface.exe!00007ff6ebcbbec6()	Unknown
 	Qt5Core.dll!0000000072e77c33()	Unknown
 	Qt5Core.dll!0000000072cc2b0f()	Unknown
 	Qt5Core.dll!0000000072cc7925()	Unknown
 	[External Code]	

I grabbed some Disassembly output let me know if you need it ( unlikely )

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2344/HighFidelity-Beta-PR10980-684001a756e65c0e9d033360ecca2106c9859ec1.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2354/HighFidelity-Beta-PR10980-3645bf04b6d5a178f2d5140f2a3e440ae542dedc.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2354/HighFidelity-Beta-PR10980-3645bf04b6d5a178f2d5140f2a3e440ae542dedc.exe

**Comment by Nex-Pro** [comment]
Retesting this.

**Comment by Nex-Pro** [comment]
For now this is a fail.

I am seeing problems with the lasers.

Example: If I am in hud/toolbar mode and I open the create panel I only get lasers onto the overlay window. When I try to laser to a cube I made the laser turns off as soon as the target leaves the overlay.


No more crashing when sweeping the teleport beam behind avatar.
Beam is smooth and fast.
Far grabbed entities now stay with the laser (nice).
Near grab works ok.




Loading laserPointerTest.js from url invoked a crash. Interface was in HMD when crash occured ( had not entered desktop mode at any time while running this ) When interface was re-started it was in desktop mode.

test mode 1. ( cyan wireframe sphere ) sometimes seems to map out the surface of the hud overlay sphere until I re-run scripts. Did not get the large red circle when doing this. Also when pulling all the way to click with the triggers the lasers turn back to blue.

test mode 2. ( yellow wireframe cube ) mostly target at ray intersection of laser sometimes limited to the hud overlay sphere radius from avatar.

test mode 3. ( purple dot ) mostly target at ray intersection of laser sometimes limited to the hud overlay sphere radius from avatar.

test mode 4. ( gnome model ) mostly target at ray intersection of laser sometimes limited to the hud overlay sphere radius from avatar.

Should any mode be intersecting avatars ? Could use some more info.


Can't get the test script to act consistently. Mostly I can see the target at the logical ray-intersection but it often seems to break. Am I doing something wrong ?

**Comment by ZappoMan** [comment]
@Nex-Pro - can you make a list of the individual problems you're seeing.... it's hard to parse out from your comment what specific things are failing. Thanks.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2368/HighFidelity-Beta-PR10980-6157f4fa55c4c1ce82657aced2c3bd75982bc283.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2368/HighFidelity-Beta-PR10980-6157f4fa55c4c1ce82657aced2c3bd75982bc283.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2373/HighFidelity-Beta-PR10980-85c3f949767babb6f667bfcc135d3b40bf6197c5.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2373/HighFidelity-Beta-PR10980-85c3f949767babb6f667bfcc135d3b40bf6197c5.exe

**Comment by Nex-Pro** [comment]
@ZappoMan I had a convo with @SamGondelman about it shortly after I posted what you read.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2431/HighFidelity-Beta-PR10980-4c558492906c3ca3744420804f90c2261c04311f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2444/HighFidelity-Beta-PR10980-f058781f49c82d97074a797ac35c24ef1b9c7a0e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2444/HighFidelity-Beta-PR10980-f058781f49c82d97074a797ac35c24ef1b9c7a0e.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2508/HighFidelity-Beta-PR10980-9bb7e6685945a9e051af3f2e0cf6dfef6880ff35.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2508/HighFidelity-Beta-PR10980-9bb7e6685945a9e051af3f2e0cf6dfef6880ff35.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2523/HighFidelity-Beta-PR10980-a641fb7a7fbe1af7a4b1b6a2cf79571203e629fe.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2523/HighFidelity-Beta-PR10980-a641fb7a7fbe1af7a4b1b6a2cf79571203e629fe.exe

**Comment by Nex-Pro** [comment]
Trying this again.

**Comment by Nex-Pro** [comment]
@SamGondelman 
Could you have someone who is responsible for the HMD tutorial take a look at these logs. I had done an express install of this build. Could not get through the HMD Tutorial.
First run: could not grab firework for second stage of tutorial. No near grab, no far grab.
Second run: As above.
Third run. Reloaded scripts managed to get to the lighter. Could not drop lighter.
Now by-passing the tutorial in order to complete testing.

**Comment by Nex-Pro** [comment]
@SamGondelman Do we need a merge from master for this? There was a protocol change yesterday. I can't go anywhere.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2533/HighFidelity-Beta-PR10980-86cc938a4741025a8e734ecfeb502d2b485d5b18.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2533/HighFidelity-Beta-PR10980-86cc938a4741025a8e734ecfeb502d2b485d5b18.dmg

**Comment by Nex-Pro** [comment]
Retesting post upstream merge

**Comment by cozza13** [comment]
Clicking around I get in a state that I can not grab the tablet and lasers do not show. Very easy to repro by playing around, teleporting, grabbing the tablet and changing domains.

**Comment by cozza13** [comment]
[hifi-log.txt](https://github.com/highfidelity/hifi/files/1232877/hifi-log.txt)

**Comment by cozza13** [comment]
- Wear the wood avatar. Go to a avatar clothing rack. try and grab some clothes. You can not.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2602/HighFidelity-Beta-PR10980-c9f4958925979aea07c2df9e68a1eea2dc02ac4e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2602/HighFidelity-Beta-PR10980-c9f4958925979aea07c2df9e68a1eea2dc02ac4e.exe
